src/main/scala/is/hail/methods/Skat.scala

```diff
@@ -0,0 +1,318 @@
+          }
+          else {
+            largeNResultOp(vArray, sigmaSq)
+          }
+          Row(k, skatStat.q, skatStat.pval, skatStat.fault)
+        }
+
+      val (skatSignature, _) = TStruct(keyName -> keyType.asInstanceOf[Type]).merge(SkatStat.schema)
+
+      new KeyTable(vds.hc, skatRDD, skatSignature, key = Array(keyName))
+    }
+
+
+    val dosages = (gs: Iterable[Genotype], n: Int) => RegressionUtils.dosages(gs, (0 until n).toArray)
+
+    (useDosages,useLargeN) match {
+      case (false, false) =>
+        val (keyedRdd, keysType, y, cov) =
+          keyedRDDSkat(vds, variantKeys, singleKey, weightExpr, yExpr, covExpr, RegressionUtils.hardCalls(_, _))
+        SkatRDDtoKeyTable(keyedRdd, keysType, y, cov, keyName, sparseResultOp)
+      case (false, true) =>
+        val (keyedRdd, keysType, y, cov) =
+          keyedRDDSkat(vds, variantKeys, singleKey, weightExpr, yExpr, covExpr, RegressionUtils.hardCalls(_, _))
+        SkatRDDtoKeyTable(keyedRdd, keysType, y, cov, keyName, largeNResultOp[SparseVector[Double]])
+      case (true, false) =>
+        val (keyedRdd, keysType, y, cov) =
+          keyedRDDSkat(vds, variantKeys, singleKey, weightExpr, yExpr, covExpr, dosages)
+        SkatRDDtoKeyTable(keyedRdd, keysType, y, cov, keyName, denseResultOp)
+      case (true, true) =>
+        val (keyedRdd, keysType, y, cov) =
+          keyedRDDSkat(vds, variantKeys, singleKey, weightExpr, yExpr, covExpr, dosages)
+        SkatRDDtoKeyTable(keyedRdd, keysType, y, cov, keyName, largeNResultOp[DenseVector[Double]])
+    }
+
+  }
+
+  def denseResultOp(st: Array[SkatTuple[DenseVector[Double]]], sigmaSq: Double): SkatStat = {
+
+    val m = st.length
+    val n = st(0).xw.size
+    val k = st(0).qtxw.size
+
+    var xwArray = new Array[Double](m * n)
+    var qtxwArray = new Array[Double](m * k)
+    var j = 0;
+    var i = 0
+
+    //copy in non-zeros to weighted genotype matrix array
+    i = 0
+    while (i < m) {
+      j = 0
+      val xwsi = st(i).xw
+      while (j < n) {
+        xwArray(i * n + j) = xwsi(j)
+        j += 1
+      }
+      i += 1
+    }
+
+    //add in non-zeros to QtGW array
+    i = 0
+    while (i < m) {
+      j = 0
+      val qtxwsi = st(i).qtxw
+      while (j < k) {
+        qtxwArray(i * k + j) = qtxwsi.data(j)
+        j += 1
+      }
+      i += 1
+    }
+
+    //compute the variance component score
+    var skatStat = 0.0
+    i = 0
+    while (i < m) {
+      skatStat += st(i).q
+      i += 1
+    }
+
+    val weightedGenotypes = new DenseMatrix[Double](n, m, xwArray)
+    val qtWeightedGenotypes = new DenseMatrix[Double](k, m, qtxwArray)
+    val GwGrammian = weightedGenotypes.t * weightedGenotypes
+    val QtGwGrammian = qtWeightedGenotypes.t * qtWeightedGenotypes
+
+    val SPG = new SkatModel(skatStat / (2 * sigmaSq))
+    SPG.computeLinearSkatStats(GwGrammian, QtGwGrammian)
+  }
+
+  def sparseResultOp(st: Array[SkatTuple[SparseVector[Double]]], sigmaSq: Double): SkatStat = {
```

Comment:
Working on abstracting it, currently having issues pattern matching on the subtypes of T <: Vector[Double]. I think it may have to do with the Java's type erasure.